api/streams.go

```diff
@@ -169,28 +178,53 @@ func (s *StreamsOf[StreamType, MsgType]) addStream(target string, component stri
 	return c, nil
 }
 
-// removeStream deletes the channel from the stream map.
-func (s *StreamsOf[StreamType, MsgType]) removeStream(target string) {
-	s.log.Debugf("Removing stream channel for target %s", target)
+// restartStream restarts a stream to the given component and starts a goroutine for sending messages from the channel to the given component
+func (s *StreamsOf[StreamType, MsgType]) restartStream(c *StreamChannelOf[StreamType, MsgType], init InitFuncOf[StreamType], opts ...grpc.DialOption) (*StreamChannelOf[StreamType, MsgType], error) {
+	var err error
 
-	s.mutex.Lock()
-	delete(s.channels, target)
-	s.mutex.Unlock()
+	// We need an init func
+	if init == nil {
+		return nil, ErrMissingInitFunc
+	}
+
+	// Initialize the stream using our init function
+	c.stream, err = init(c.target, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("could not init stream: %w", err)
+	}
+
+	// Revive the stream
+	c.dead = false
+
+	s.log.Infof("Re-Established stream to %s (%s)", c.component, c.target)
+
+	// Start go routine for receiving messages from the stream (especially relevant for bi-directional streams).
+	go c.recvLoop(s)
+
+	// Start go routine for sending messages from the channel to the stream
+	go c.sendLoop(s)
+
+	return c, nil
 }
 
 // sendLoop continuously fetches new messages from the channel inside c and sends them to the appropriate stream.
 func (c *StreamChannelOf[StreamType, MsgType]) sendLoop(s *StreamsOf[StreamType, MsgType]) {
 	var err error
 
-	// Fetch new messages from channel (this will block)
+	// Fetch new (or queued old) messages from the channel. This will block.
 	for m := range c.channel {
 		// Try to send the message in our stream
 		err = c.stream.SendMsg(m)
 		if errors.Is(err, io.EOF) {
 			s.log.Infof("Stream to %s (%s) closed with EOF", c.component, c.target)
 
-			// Remove the stream from our map and end this goroutine
-			s.removeStream(c.target)
+			// Declare the stream as dead
+			c.dead = true
+
+			// Put the message back on the channel, so that it does not get lost
+			go func() {
+				c.channel <- m
+			}()
```

Comment:
We could also try to restart the stream here too.

Reply:
> We could also try to restart the stream here too.

Restarting the stream in `sendLoop` is problematic because the `init` might fail and we cannot return an error in the sendLoop.